src/storage/v2/vertex_accessor.cpp

```diff
@@ -34,31 +36,39 @@ std::pair<bool, bool> IsVisible(Vertex *vertex, Transaction *transaction, View v
     deleted = vertex->deleted;
     delta = vertex->delta;
   }
-  ApplyDeltasForRead(transaction, delta, view, [&](const Delta &delta) {
-    switch (delta.action) {
-      case Delta::Action::ADD_LABEL:
-      case Delta::Action::REMOVE_LABEL:
-      case Delta::Action::SET_PROPERTY:
-      case Delta::Action::ADD_IN_EDGE:
-      case Delta::Action::ADD_OUT_EDGE:
-      case Delta::Action::REMOVE_IN_EDGE:
-      case Delta::Action::REMOVE_OUT_EDGE:
-        break;
-      case Delta::Action::RECREATE_OBJECT: {
-        deleted = false;
-        break;
-      }
-      case Delta::Action::DELETE_DESERIALIZED_OBJECT:
-      case Delta::Action::DELETE_OBJECT: {
-        exists = false;
-        break;
-      }
+
+  // Checking cache has a cost, only do it if we have any deltas
+  // if we have no deltas then what we already have from the vertex is correct.
+  if (delta && transaction->isolation_level != IsolationLevel::READ_UNCOMMITTED) {
+    // IsolationLevel::READ_COMMITTED would be tricky to propagate invalidation to
+    // so for now only cache for IsolationLevel::SNAPSHOT_ISOLATION
+    auto const useCache = transaction->isolation_level == IsolationLevel::SNAPSHOT_ISOLATION;
+
+    if (useCache) {
+      auto const &cache = transaction->manyDeltasCache;
+      auto existsRes = cache.GetExists(view, vertex);
+      auto deletedRes = cache.GetDeleted(view, vertex);
+      if (existsRes && deletedRes) return {*existsRes, *deletedRes};
     }
-  });
+
+    auto const n_processed = ApplyDeltasForRead(transaction, delta, view, [&](const Delta &delta) {
+      // clang-format off
+      DeltaDispatch(delta, utils::ChainedOverloaded{
+          Deleted_ActionMethod(deleted),
+          Exists_ActionMethod(exists)
+      });
+      // clang-format on
+    });
+
+    if (useCache && n_processed >= FLAGS_delta_chain_cache_threashold) {
+      auto &cache = transaction->manyDeltasCache;
+      cache.StoreExists(view, vertex, exists);
+      cache.StoreDeleted(view, vertex, deleted);
```

Comment:
maybe it does not make sense if I think of it because how would you know if you need to update something. perhaps use a `std::optional` for every time. idk, I'm just discussing what's better at this point although everything looks clear to me